sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ToNumberParser.scala

```diff
@@ -0,0 +1,517 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.util
+
+import scala.collection.mutable
+
+import org.apache.spark.sql.catalyst.analysis.TypeCheckResult
+import org.apache.spark.sql.errors.QueryExecutionErrors
+import org.apache.spark.sql.types.{Decimal, DecimalType}
+import org.apache.spark.unsafe.types.UTF8String
+
+// This object contains some definitions of characters and tokens for the parser below.
+object ToNumberParser {
+  final val ANGLE_BRACKET_CLOSE = '>'
+  final val ANGLE_BRACKET_OPEN = '<'
+  final val COMMA_LETTER = 'G'
+  final val COMMA_SIGN = ','
+  final val DOLLAR_LETTER = 'L'
+  final val DOLLAR_SIGN = '$'
+  final val MINUS_SIGN = '-'
+  final val NINE_DIGIT = '9'
+  final val OPTIONAL_PLUS_OR_MINUS_LETTER = 'S'
+  final val PLUS_SIGN = '+'
+  final val POINT_LETTER = 'D'
+  final val POINT_SIGN = '.'
+  final val POUND_SIGN = '#'
+  final val ZERO_DIGIT = '0'
+
+  final val OPTIONAL_MINUS_STRING = "MI"
+  final val WRAPPING_ANGLE_BRACKETS_TO_NEGATIVE_NUMBER = "PR"
+
+  final val OPTIONAL_MINUS_STRING_START = 'M'
+  final val OPTIONAL_MINUS_STRING_END = 'I'
+  final val WRAPPING_ANGLE_BRACKETS_TO_NEGATIVE_NUMBER_START = 'P'
+  final val WRAPPING_ANGLE_BRACKETS_TO_NEGATIVE_NUMBER_END = 'R'
+
+  // This class represents one or more characters that we expect to be present in the input string
+  // based on the format string.
+  abstract class InputToken()
+  // Represents some number of digits (0-9).
+  abstract class Digits extends InputToken
+  // Represents exactly 'num' digits (0-9).
+  case class ExactlyAsManyDigits(num: Int) extends Digits
+  // Represents at most 'num' digits (0-9).
+  case class AtMostAsManyDigits(num: Int) extends Digits
+  // Represents one decimal point (.).
+  case class DecimalPoint() extends InputToken
+  // Represents one thousands separator (,).
+  case class ThousandsSeparator() extends InputToken
+  // Represents one or more groups of digits (0-9) with thousands separators (,) between each group.
+  case class DigitGroups(tokens: Seq[InputToken]) extends InputToken
```

Comment:
```suggestion
  case class DigitGroups(tokens: Seq[Digits]) extends InputToken
```

Reply:
Rejected this suggestion because the `DigitGroups` can actually contain children comprising either `Digits` or `ThousandsSeparators`.